rust/src/mainthread.rs

```diff
@@ -0,0 +1,204 @@
+//! This module provides two ways to execute "jobs":
+//!
+//! 1. On the Binary Ninja main thread (the UI event thread when running in the GUI application):
+//!    * [execute_on_main_thread]
+//!    * [execute_on_main_thread_and_wait]
+//! 2. On a worker thread
+//!
+//! Any manipulation of the GUI should be performed on the main thread, but any
+//! non-GUI work is generally better to be performed using a worker. This is
+//! especially true for any longer-running work, as the user interface will
+//! be unable to update itself while a job is executing on the main thread.
+//!
+//! There are three worker queues, in order of decreasing priority:
+//!
+//!    1. The Interactive Queue ([worker_interactive_enqueue])
+//!    2. The Priority Queue ([worker_priority_enqueue])
+//!    3. The Worker Queue ([worker_enqueue])
+//!
+//! All of these queues are serviced by the same pool of worker threads. The
+//! difference between the queues is basically one of priority: one queue must
+//! be empty of jobs before a worker thread will execute a job from a lower
+//! priority queue.
+//!
+//! The default maximum number of concurrent worker threads is controlled by the
+//! `analysis.limits.workerThreadCount` setting but can be adjusted at runtime via
+//! [set_worker_thread_count].
+//!
+//! The worker threads are native threads, managed by the Binary Ninja core. If
+//! more control over the thread is required, consider using the
+//! [crate::backgroundtask::BackgroundTask] class.
+
+use core::{ffi, mem, ptr};
+
+use binaryninjacore_sys::*;
+
+use crate::string::BnStrCompatible;
+
+#[repr(transparent)]
+pub struct MainThreadAction {
+    handle: ptr::NonNull<BNMainThreadAction>,
+}
+
+impl Clone for MainThreadAction {
+    fn clone(&self) -> Self {
+        let result = unsafe { BNNewMainThreadActionReference(self.as_raw()) };
+        unsafe { Self::from_raw(ptr::NonNull::new(result).unwrap()) }
+    }
+}
+
+impl Drop for MainThreadAction {
+    fn drop(&mut self) {
+        unsafe { BNFreeMainThreadAction(self.as_raw()) }
+    }
+}
+
+impl MainThreadAction {
+    pub(crate) unsafe fn from_raw(handle: ptr::NonNull<BNMainThreadAction>) -> Self {
+        Self { handle }
+    }
+
+    pub(crate) unsafe fn ref_from_raw(handle: &*mut BNMainThreadAction) -> &Self {
+        assert!(!handle.is_null());
+        mem::transmute(handle)
+    }
+
+    #[allow(clippy::mut_from_ref)]
+    pub(crate) unsafe fn as_raw(&self) -> &mut BNMainThreadAction {
+        &mut *self.handle.as_ptr()
+    }
```

Comment:
```suggestion
    pub(crate) fn as_raw(&self) -> *mut BNMainThreadAction {
        self.handle.as_ptr()
    }
```

This function isn't `pub` and passing the raw `*mut BNMainThreadAction` across FFI already requires unsafe, so this function should just be a wrapper around `NonNull::as_ptr`. As far as I can tell, every use of this function coerces the return type from `&mut` to `*mut`.

Reply:
Similar to above, the `*mut T: Copy` is avoided and lifetime is enforced to avoid citations like #5496.

Although the pointer is usually immediately converted into a pointer, voiding the effort of enforcing the lifetime. In any case, this is a internal function and it could be fixed without changing the API.